File: minStack/minStack2.cpp
#include <iostream>
#include <stack>
using namespace std;

class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {
        minV = INT_MAX;
    }
    
    void push(int x) {
        if (x <= minV) {
            s.push(minV);
            minV = x;
        }
        s.push(x);
    }
    
    void pop() {
        if (!s.empty()) {
            if (s.top() == minV) {
                s.pop();
                minV = s.top();
            }
            s.pop();
        }
    }
    
    int top() {
        return s.top();
    }
    
    int getMin() {
        return minV;
    }
    int minV;
    stack<int> s;
};

int main()
{
    MinStack obj;
    int i = 1;
    while (i < 10)
        obj.push(i++);
    obj.pop();
    int param_3 = obj.top();
    bool param_4 = obj.getMin();
    cout << param_3 << " " << param_4 << endl;
    return 0;
}

File: 155.minStack/minStack2.java
class MinStack {

    /** initialize your data structure here. */
    public MinStack() {
        s = new Stack<Integer>();
        minV = 0;
    }
    
    public void push(int x) {
        if (s.isEmpty() || x <= minV) {
            s.push(minV);
            minV = x;
        }
        s.push(x);
    }
    
    public void pop() {
        if (!s.isEmpty()) {
            if (s.peek().equals(minV)) {
                s.pop();
                minV = s.pop();
            } else {
                s.pop();
            }
        }
        //printS(minS);
    }
    
    public int top() {
        int res = -1;
        if (!s.isEmpty())
            res = s.peek();
        return res;
    }
    public void printS(Stack<Integer> s) {
        Stack<Integer> ss = (Stack)s.clone();
        while (!ss.isEmpty()) {
            System.out.print(ss.pop() + " ");
        }
        System.out.println();
    }
    public int getMin() {
        int res = -1;
        if (!s.isEmpty())
            res = minV;
        return res;
    }
    Stack<Integer> s;
    int minV;
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
